### ISSUE TITLE
SizeOfReturn doesn't work on Library Data Type Methods

### ISSUE THREAD

**Issue by piboistudios** [Aurel300/ammer#32]
The workaround is to store a reference to the pointer globally in C so it can be used statically, and move the Haxe method from the Pointer type to the Library type as a static function.

**Comment by Aurel300** [comment]
I'll investigate, but I am not sure about your workaround. Instead of:

```
// in the data type
public function foo(_:ammer.ffi.This, x:Int, y:Int):Int;
```

You can always just have:

```
// in the library
public static function foo(_:TheDataType, x:Int, y:Int):Int;
```

**Comment by piboistudios** [comment]
Oh, I wasn't sure if that worked. That's awesome!

**Comment by piboistudios** [comment]
@Aurel300  Not sure if you want me to open another issue for this, but somewhat related, different Library Data Types can't share the same method names.
Example:
```haxe
@:ammer.nativePrefix("critical_section_")
class CriticalSection extends Pointer<"critical_section_t", SynchLib> {
  public function enter(_:This):Void;
  public function leave(_:This):Void;
  public function delete(_:This):Void;
}
@:ammer.nativePrefix("synch_barrier_")
class SynchronizationBarrier extends Pointer<"barrier_t", SynchLib> {
  public function enter(_:This):Void;
  public function delete(_:This):Void;
}
```
Won't compile because all the functions are hoisted to the static `Library` type; maybe namespace them using the child name? i.e. `'${libPack}.${libType}.${pointerClassName}_${method_name}'`; the above example would then add the methods:
`synch.SynchLib.SynchronizationBarrier_enter`, `synch.SynchLib.SynchronizationBarrier_delete`, `synch.SynchLib.CriticalSection_enter`, and `synch.SynchLib.SynchronizationBarrier_leave`, and `synch.SynchLib.SynchronizationBarrier_delete` (right now this results in duplicate method name issues).

Or just use the fully qualified native name (i.e. `nativePrefix` + `methodName`)

**Comment by piboistudios** [comment]
You also can't extend Library Data Types; this is less useful, but still has some relevance to addressing the way C/C++ expects certain usage of certain pointers/handles based on where they come from.

For example, Windows uses `HANDLE` for a great many different things and if you use that type incorrectly, you get an error (i.e., if you create a HANDLE for an Event and try to Release it like a Mutex, error, presumably). It would be nice to be able to inherit other `Pointer` types for this reason (then you can have different Haxe types represent the same Opaque and only expose the needed methods on the corresponding Haxe types), but I imagine that may be far more complex than the workaround (simply adding a second layer of Haxe-types ontop of Natives to enforce these type restrictions; which is somewhat unavoidable anyways if you want a user-friendly/Haxey looking API, hence, this is a bit less useful)

**Comment by Aurel300** [comment]
> Not sure if you want me to open another issue for this, but somewhat related, different Library Data Types can't share the same method names.

No need for a separate issue for this. And yes, good point; and yes, it can be fixed by prefixing the generated library functions. :)

> You also can't extend Library Data Types; this is less useful, but still has some relevance to addressing the way C/C++ expects certain usage of certain pointers/handles based on where they come from.

I hadn't though of that much yet, but inheritance might indeed be a good way to model pointers to structs like `SDL_event` which have a "base" layout + essentially a `union` of possible sub-types. I'll keep that in mind. I will try to release an `ammer` definition for SDL2 at some point, so I will see how inheritance might help there.

**Comment by Aurel300** [comment]
> Not sure if you want me to open another issue for this, but somewhat related, different Library Data Types can't share the same method names.

This is now fixed with 89f9e64.
